docs/internal/types.md

```diff
@@ -0,0 +1,287 @@
+There is no corresponding `unfreeze` operation to consume an `imm` capability and provide any kind of mutable capability.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific monotonic semantics that change the set of operations that are permitted with the resulting capability.
+
+## Module / Class model
+
+TODO
+
+## Classes and Interfaces
+
+Classes and interfaces have a similar structure:
+ * Fields: elements of a specific type that have a runtime space allocated.
+ * Methods: functions that belong to a type.
+
+The declaration is trivial:
+```ts
+// Helps receivers know they can call a function named increment
+interface Example
+{
+  // All derived classes will have the common field
+  var common : I64;
+
+  // Classes that implement this interface must implement this method and provide the field
+  // Classes that inherit from this interface must implement this method but get the field from the base class
+  increment();
+};
+
+// This class inherits, so it gets the field `common` from the interface
+class MyType : Example
+{
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 1;
+  }
+};
+
+// This class implements the interface, so it must provide the `common` field
+class OtherType {
+  // Must provide the field
+  var common : I64;
+
+  // Implements `increment`
+  increment()
+  {
+    self.common += 2;
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class or interface and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later.
+
+They can also have a keyword `embed`, which means ...
+
+### Methods
+
+Methods are functions that belong to a class or interface which can have arguments and a return type.
+
+Methods can be called statically (via `Class::method()`) or dynamically (via `object.method()`).
+On the static case, the `self` pointer is a descriptor for the type, while on the dynamic case, it's a descriptor to the instance (and its vtable, type info, etc).
+
+### Constructors
+
+Verona distinguishes memory allocation from object construction.
+Verona's built-in `new` function fully constructs an object and so must be called with values for all of the fields in the object.
+It may be used only from static methods on the class and so is usually used by factory methods.
+By convention, factory methods are called `create()` and there is syntactic sugar to allow `T()` to implicitly call `T::create()`.
+
+Example:
+```ts
+class Capibara
+{
+  let name : String & imm;
+  var weight : F32; // Kgs
+
+  // Create a baby capibara
+  newborn(name : String) : Capibara
```

Comment:
Maybe rename this `create`?  It seems odd to say 'by convention these things are called `create`' and then not call this one `create`.  I'm not 100% sure on this, because I quite like the fact that it shows that factory methods don't *have* to be named `create`.

Reply:
Yeah, this is why I wanted it to not be `create`. I can add another method called `create` with no arguments for completeness.